### ISSUE TITLE
Potential regression related to response state handling/committing after update from 9.4.20 to 9.4.22

### ISSUE THREAD

**Issue by e-hubert** [eclipse/jetty.project#4303]
Although we did not face any issues during both automated regression testing (all our integration, stress and performance tests passed) as well as manual testing of our platform services after upgrading from Jetty 9.4.20 to 9.4.22 immediately after the first frontend service instances were put into production, we sporadically faced the following two types of likely closely related issues while serving simple requests (e.g. GET requests for javascript or image resource files).

```
java.io.IOException: cannot sendContent(), output already committed
	at org.eclipse.jetty.server.HttpOutput.sendContent(HttpOutput.java:1048)
	at org.eclipse.jetty.server.HttpOutput.sendContent(HttpOutput.java:958)
	at org.eclipse.jetty.server.ResourceService.sendData(ResourceService.java:725)
	at org.eclipse.jetty.server.ResourceService.doGet(ResourceService.java:294)
	at org.eclipse.jetty.servlet.DefaultServlet.doGet(DefaultServlet.java:454)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:687)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder$NotAsyncServlet.service(ServletHolder.java:1401)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:760)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1617)
...
```

```
java.lang.IllegalStateException: Response is COMPLETED
	at org.eclipse.jetty.server.HttpChannelState.sendError(HttpChannelState.java:901)
	at org.eclipse.jetty.server.Response.sendError(Response.java:430)
	at org.eclipse.jetty.server.Response.sendError(Response.java:401)
...
```
While we were sporadically facing the issue in production we tried to force it by manually sending the exact same requests, but those succeeded. According to our logs also the majority of all live requests for the same URLs  succeeded. It somehow looks like some race/concurrency issue.

I know this is not much (maybe too less) information to report a bug, but unfortunately at the moment I do not have more information available. Do to uncertaincies about the business impact we immediately rolled back to 9.4.20 and those errors immediately disappeared.

I briefly looked over the changes between the Jetty versions and this PR https://github.com/eclipse/jetty.project/pull/3912 caught my attention, but I did not immediately spot somethin (the changeset is also quite large). The area around HttpChannelState/OutputState might be relevant here.

Unfortunately, so far I had no time to investigate any further. Of course it would be great to have an isolated reproducer, but I don't have one.

Do you maybe have some suggestion how to pinpoint the root cause? I cannot promise, but I try to find some time later this week to provide more information.

**Comment by joakime** [comment]
Based on this scant information I think your have 2 issues.

1. The `DefaultServlet` / `ResourceService` one is likely handled by PR #4294 
2. The `sendError()` issue is definitely related to #3912 but there's not enough of a stacktrace to make any statement on cause.  
    It's saying it's unable to send an error because the response is already completed.
    That could be legitimate, as a result of a bad usage of `sendError()` (attempting to use `HttpServletResponse.sendError()` at the wrong time.
    That could also be a side effect of what PR #4294 is fixing (if the request that is causing the Exception occurs after another one on a persistent connection that triggered the issue that PR #4294 fixes)
   
If you can reproduce the issue, I would recommend building branch `jetty-9.4.x-4161-Write-LifeCycleViolation` and trying it out with your reproducer.

**Comment by joakime** [comment]
Whoop, never mind, branch `jetty-9.4.x-4161-Write-LifeCycleViolation` has just been merged into mainline `jetty-9.4.x` branch a moment ago.

**Comment by e-hubert** [comment]
Thanks for the quick and valuable response @joakime!

The first issue seems to be by far more important than the second one (sendError), which may indeed also be a wrong usage (I have to verify this in detail). It looks like we call response.sendError(404) in a StripesResolution passed to a Stripes DefaultViewActionBean in an overridden implementation of net.sourceforge.stripes.controller.NameBasedActionResolver#handleActionBeanNotFound.
I noticed the second error occurred only once on a single server after the update, while the first one occurred on many systems a few dozen times.
In case I should be able to reproduce it locally, I can of cause give a 9.4.23-SNAPSHOT build from https://oss.sonatype.org/content/repositories/jetty-snapshots a try. At this time there will likely be a timestamped snapshot containing this merge available, if not I'll build it locally. Thanks again!

**Comment by e-hubert** [comment]
Unfortunately I had no time to work on a reproduction today, but I just want to add something I missed initially. Internal clients of our affected services running on Jetty accessing Jetty via an http server as a reverse proxy received an identical number of 502 responses as a result of proxy timeouts giving me more evidence it wasn't a second attempt to commit something, but rather a wrong internal state resulting in nothing send to the client.

**Comment by joakime** [comment]
The current 9.4.23-SNAPSHOT builds on https://oss.sonatype.org/content/repositories/jetty-snapshots are ready for you to test against.

**Comment by e-hubert-opti** [comment]
Seems I can confirm the issue to be resolved with 9.4.23. Today we also updated the first productive instance to 9.4.23 and so far no problem. Tomorrow we plan continue updating more servers straight to 9.4.24. So I intent to close the ticket tomorrow, but just wanted to provide feedback as soon as possible - sorry I did not manage to get back earlier.

**Comment by e-hubert-opti** [comment]
As said - I close this now, because we are running in production with at least one complete service cluster on 9.4.24 for about an hour without any issues, so I consider this one fixed since 9.4.23. Thanks a lot!
